electrum/lnpeer.py

```diff
@@ -220,12 +220,100 @@ def process_message(self, message):
+        # otherwise:
+        #     MUST fail the channel referred to by channel_id, if that channel is with the sending node.
+        elif channel_id in self.temp_id_to_id:
+            channel_id = self.temp_id_to_id[channel_id]
+        elif channel_id in self.channels:
+            await self.lnworker.force_close_channel(channel_id)
+        else:
+            channel_id = None
+        if channel_id is not None:
+            self.ordered_message_queues[channel_id].put_nowait((None, {'error': payload['data']}))
+            raise GracefulDisconnect
+
+    async def send_warning(self, channel_id: bytes, message: str = None, *, close_connection=True):
+        """Sends a warning and disconnects if close_connection.
+
+        Note:
+        * channel_id is the temporary channel id when the channel id is not yet available
+
+        A sending node:
+        MAY set channel_id to all zero if the warning is not related to a specific channel.
+
+        when failure was caused by an invalid signature check:
+        * SHOULD include the raw, hex-encoded transaction in reply to a funding_created,
+          funding_signed, closing_signed, or commitment_signed message.
+        """
+        assert isinstance(channel_id, bytes)
+        encoded_data = b'' if not message else message.encode('ascii')
+        self.send_message('warning', channel_id=channel_id, data=encoded_data, len=len(encoded_data))
+        if close_connection:
+            raise GracefulDisconnect
+
+    async def send_error(self, channel_id: bytes, message: str = None, *, force_close_channel=False):
+        """Sends an error message and force closes the channel.
+
+        Note:
+        * channel_id is the temporary channel id when the channel id is not yet available
+
+        A sending node:
+        * SHOULD send error for protocol violations or internal errors that make channels
+          unusable or that make further communication unusable.
+        * SHOULD send error with the unknown channel_id in reply to messages of type
+          32-255 related to unknown channels.
+        * MUST fail the channel(s) referred to by the error message.
+        * MAY set channel_id to all zero to indicate all channels.
+
+        when failure was caused by an invalid signature check:
+        * SHOULD include the raw, hex-encoded transaction in reply to a funding_created,
+          funding_signed, closing_signed, or commitment_signed message.
+        """
+        assert isinstance(channel_id, bytes)
+        encoded_data = b'' if not message else message.encode('ascii')
+        self.send_message('error', channel_id=channel_id, data=encoded_data, len=len(encoded_data))
+        # channel_id of zero means that the error refers to all channels
+        if channel_id == bytes(32):
+            for channel_id in self.channels:
+                await self.lnworker.force_close_channel(channel_id)
```

Comment:
shouldn't this also be guarded with `force_close_channel`?

Reply:
I have channel force closing in both scenarios. The case with all-zero bits is a bit dangerous.

Reply:
Sorry, I'm not sure I understand.
My question is this: What is the expected behaviour of calling `peer.send_error(bytes(32), force_close_channel=False)`?
Right now, AFAICT, it would fclose the channels.

Reply:
Ah sorry, I misunderstood, it is guarded now. The spec says we must force close upon sending an error (individual or all channels), but I would keep the `force_close_channel` parameter nevertheless.

Reply:
Another question is if we really should allow force closing of all channels if we receive the zero `channel_id`, it looks dangerous.

Reply:
The BOLT says we must fail channels with the sending node. If the sending node wants all their channels closed, there is not much we can do... Are you concerned about malicious behaviour by the remote, or programming bugs in the remote's implementation, or something else?

Re malicious behaviour, I was thinking we might introduce `Peer.force_close_channel` and `Peer.try_force_closing` as wrappers for `LNWorker.force_close_channel` and `LNWorker.try_force_closing`, that would enforce that the `channel_id` corresponds to a channel with this Peer. This does not have to be a part of this PR though.

Reply:
> The BOLT says we must fail channels with the sending node. If the sending node wants all their channels closed, there is not much we can do... Are you concerned about malicious behaviour by the remote, or programming bugs in the remote's implementation, or something else?

Yes, mostly about something that could trigger another node sending the zero message (or all nodes in the network across multiple implementations :P).

>Re malicious behaviour, I was thinking we might introduce Peer.force_close_channel and Peer.try_force_closing as wrappers for LNWorker.force_close_channel and LNWorker.try_force_closing, that would enforce that the channel_id corresponds to a channel with this Peer. This does not have to be a part of this PR though.

Sounds useful, added the wrapper and used it instead.

Reply:
Should we also loop over the temporary channel ids?